File: src/go/problems/MaxPathSum_test.go
// https://leetcode.com/problems/binary-tree-maximum-path-sum/

package problems

import (
	"testing"
)

func TestMaxPathSum(t *testing.T) {
	var node = TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	
	result := MaxPathSum(&node)
	t.Log(result)
}

var result int

func MaxPathSum(root *TreeNode) int {
    result = root.Val
	Dfs(root)

	return result
}

func Dfs(node *TreeNode) int {
	if node == nil {
		return 0
	}

	left := Dfs(node.Left)
	right := Dfs(node.Right)

	nodeMax := max2(node.Val + left, node.Val + right)
	nodeMax = max2(nodeMax, node.Val)

	result = max2(result, nodeMax)
	result = max2(result, left + right + node.Val)

	return nodeMax
}

func max2(left int, right int) int {
	if left > right {
		return left
	}

	return right
}

File: src/python/problems/max_path_sum_test.py
# https://leetcode.com/problems/binary-tree-maximum-path-sum/

from typing import Optional
from models.tree_node import TreeNode
from core.problem_base import *

class MaxPathSum(ProblemBase):
    def Solution(self, root: Optional[TreeNode]) -> int:
        result = [root.val]

        def Dfs(node) -> int:
            if not node:
                return 0

            left = Dfs(node.left)
            right = Dfs(node.right)

            nodeMax = max(node.val + left, node.val + right)
            nodeMax = max(nodeMax, node.val)

            result[0] = max(result[0], nodeMax)
            result[0] = max(result[0], left + right + node.val)

            return nodeMax

        Dfs(root)

        return result[0]

if __name__ == '__main__':
    TestGen(MaxPathSum) \
        .Add(lambda tc: tc.ParamTreeNode([1,2,3]).Result(6)) \
        .Add(lambda tc: tc.ParamTreeNode([-10,9,20,None,None,15,7]).Result(42)) \
        .Run()
